server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourcePhase.java

```diff
@@ -47,7 +47,8 @@ public StoredFieldsSpec storedFieldsSpec() {
             public void process(HitContext hitContext) {
                 String index = fetchContext.getIndexName();
                 if (fetchContext.getSearchExecutionContext().isSourceEnabled() == false) {
-                    if (fetchSourceContext.hasFilter()) {
+                    if (fetchSourceContext.hasFilter()
+                        || (fetchSourceContext.includeVectors() != null && fetchSourceContext.includeVectors() == false)) {
```

Comment:
It may seem wrong to throw here if `fetchSourceContext.includeVectors() == false`, but it's actually more technically correct because we act on this value by filtering _out_ vectors (i.e. adding vector fields to the list of excluded fields). Thus, when `fetchSourceContext.includeVectors() == true`, we don't take any action.

Throwing only on `fetchSourceContext.includeVectors() == false` also has the important side-effect of fixing some YAML test failures. In particular, [reindex](https://github.com/Mikep86/elasticsearch/blob/e032abfdbb2d6e57eabc81de116f961271b959ca/modules/reindex/src/yamlRestTest/resources/rest-api-spec/test/reindex/20_validation.yml#L254-L254) and [update-by-query](https://github.com/Mikep86/elasticsearch/blob/e032abfdbb2d6e57eabc81de116f961271b959ca/modules/reindex/src/yamlRestTest/resources/rest-api-spec/test/update_by_query/20_validation.yml#L102-L102) validation tests expect an exception [thrown from here](https://github.com/Mikep86/elasticsearch/blob/e032abfdbb2d6e57eabc81de116f961271b959ca/modules/reindex/src/main/java/org/elasticsearch/reindex/AbstractAsyncBulkByScrollAction.java#L265-L265) when _source is disabled in the mappings. However, if we throw on `fetchSourceContext.includeVectors() == true`, that behavior changes and sometimes we throw from here when _source is disabled in the mappings, breaking those tests. You may think we can fix this by updating the tests, but unfortunately it's not that simple because we do not always throw from here when source is disabled in the mappings. Take the following scenario:

- source is disabled in the mappings
- The user executes a reindex where `include_vectors` is set to `false`

In this case, we will not throw here and instead throw in `AbstractAsyncBulkByScrollAction`, creating inconsistent behavior for the same error case. This could be very confusing for a user.

Reply:
> source is disabled in the mappings

If `_source` is disabled in mappings shouldn't that supersede whether vectors should be included in `_source`? I feel like there's a conflicting parameter validation that we could do in request constructors here?  

> The user executes a reindex where include_vectors is set to false (yes, this is possible)

Wouldn't the correct action be to disallow this when constructing the `reindex` command and return a 400-level error if a user attempts a reindex without enabling `_source` or vectors in `_source`? That would also enable us to return a "friendlier" error telling users "you can't make that request"

Reply:
> If _source is disabled in mappings shouldn't that supersede whether vectors should be included in _source? I feel like there's a conflicting parameter validation that we could do in request constructors here?

We don't have the mappings at the time the request is constructed and thus we cannot check if `_source` is disabled in the mappings at that time. This is the first place we have access to both the mappings and the request.

Reply:
We can simplify this logic and preserve existing error handling characteristics by using `SourceFilter#isEmpty` here instead. It also aligns with @jimczi's [request to remove any dependence on `filterVectorFields`](https://github.com/elastic/elasticsearch/pull/113418#discussion_r1781868225) (this check is essentially a subset of what is checked in that method). Overall looks like a tidy solution. @kderusso @carlosdelest WDYT?

Reply:
@Mikep86 I think that's a good solution, thanks!